File: bundle/giapi-gmp/gmp-statusdb/src/edu/gemini/aspen/gmp/statusdb/osgi/Activator.java
package edu.gemini.aspen.gmp.statusdb.osgi;

import org.osgi.framework.BundleActivator;
import org.osgi.framework.BundleContext;
import org.osgi.framework.ServiceRegistration;
import edu.gemini.aspen.gmp.statusdb.StatusDatabase;
import edu.gemini.aspen.gmp.status.api.StatusHandler;
import edu.gemini.aspen.gmp.status.api.StatusDatabaseService;

/**
 * Activator class for the Status Database bundle
 */
public class Activator implements BundleActivator {

    private StatusProcessorTracker _tracker;
    private StatusDatabase _database;

    private ServiceRegistration _shRegistration;
    private ServiceRegistration _dbRegistration;

    public void start(BundleContext bundleContext) throws Exception {

        _database = new StatusDatabase();
        _database.start();

        //advertise the Status Database into OSGi
        _shRegistration = bundleContext.registerService(
                StatusHandler.class.getName(),
                _database, null);

        //and advertise it as a Database Service as well.
        _dbRegistration = bundleContext.registerService(
                StatusDatabaseService.class.getName(),
                _database, null);

        //watch for status processors to be registered in this database
        _tracker = new StatusProcessorTracker(bundleContext, _database);
        _tracker.open();

    }

    public void stop(BundleContext bundleContext) throws Exception {

        _tracker.close();
        _tracker = null;

        _database.shutdown();

        _database = null;

        _shRegistration.unregister();
        _dbRegistration.unregister();


    }
}

File: gmp-tcs-offset/src/main/scala/edu/gemini/aspen/gmp/tcsoffset/osgi/Activator.scala
package edu.gemini.aspen.gmp.tcsoffset.osgi

import java.util
import edu.gemini.epics.{EpicsObserver, EpicsReader, EpicsWriter}
import edu.gemini.util.osgi.Tracker
import org.osgi.framework.{BundleActivator, BundleContext, ServiceRegistration}
import org.osgi.service.cm.ManagedServiceFactory
import org.osgi.util.tracker.ServiceTracker

import java.util.logging.Logger
import edu.gemini.aspen.gmp.tcsoffset.model.TcsOffsetComponent

/**
 * Activator class is the responsible for implementing the start and stop
 * of the gmp-tcs-offset bundle.
 */
class Activator extends BundleActivator {
  private val logger = Logger.getLogger(this.getClass.getName)
  var trackerTop: Option[ServiceTracker[EpicsWriter, _]] = None
  var psService: Option[ServiceRegistration[_]] = None

  override def start(context: BundleContext): Unit = {
    logger.fine("TCS Offset bundle Starting")
    trackerTop = Option(Tracker.track[EpicsWriter, EpicsObserver, TcsOffsetComponentFactory](context) { (ew, eo) =>
    val ps = new TcsOffsetComponentFactory(context, ew, eo)
    val props: util.Hashtable[String, String] = new util.Hashtable()
    props.put("service.pid", classOf[TcsOffsetComponent].getName)
    psService = Option(context.registerService(classOf[ManagedServiceFactory].getName, ps, props))
    ps
    } { p =>
        p.stopServices()
        psService.foreach(_.unregister())
        psService = None
    })
    trackerTop.foreach(_.open)
  }

  override def stop(context: BundleContext): Unit = {
    trackerTop.foreach(_.close())
    trackerTop = None
  }
}
